bestie/server/testresult.go

```diff
@@ -0,0 +1,220 @@
+	stream := bazelStream{
+		buildId:      streamId.GetBuildId(),
+		invocationId: streamId.GetInvocationId(),
+		run:          strconv.Itoa(int(bazelBuildEvent.GetId().GetTestResult().GetRun())),
+		testName:     bazelBuildEvent.GetId().GetTestResult().GetLabel(),
+	}
+	// Calculate a SHA256 hash using the following fields to uniquely identify this stream.
+	stream.invocationSha = deriveInvocationSha([]string{stream.invocationId, stream.buildId, stream.run})
+	return stream
+}
+
+// Handle metrics extraction from the TestResult event.
+func handleTestResultEvent(bazelBuildEvent bes.BuildEvent, streamId *build.StreamId) error {
+	stream := identifyStream(bazelBuildEvent, streamId)
+	m := bazelBuildEvent.GetTestResult()
+	if m == nil {
+		return fmt.Errorf("Error extracting TestResult data from event message")
+	}
+
+	fmt.Printf("TestResult for %s: %s\n", stream.testName, m.GetStatus())
+	fmt.Printf("  run: %s\n", stream.run)
+	fmt.Printf("  buildId: %s\n", stream.buildId)
+	fmt.Printf("  invocationId: %s\n", stream.invocationId)
+	fmt.Printf("  invocationSha: %s\n\n", stream.invocationSha)
+
+	outputFiles := m.GetTestActionOutput()
+	var ofname, ofuri string
+	found := false
+	for _, of := range outputFiles {
+		ofname = of.GetName()
+		ofuri = of.GetUri()
+		if strings.HasSuffix(ofname, "outputs.zip") {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil
+	}
+
+	// Strip off any file:// prefix from the URI to access the local file system path.
+	filePrefix := "file://"
+	if strings.HasPrefix(ofuri, filePrefix) {
+		ofuri = ofuri[len(filePrefix):]
+	}
+
+	// Process test metrics output file(s).
+	// Each output file contains a single (potentially large) protobuf message.
+	if err := extractZippedFiles(stream, ofuri); err != nil {
+		return fmt.Errorf("Error processing %s file: %s", ofname, err)
+	}
+
+	return nil
+}
+
+// Look for and return a []byte slice with its contents.
+func extractZippedFiles(stream bazelStream, zipFile string) error {
+	metricsFileRE := regexp.MustCompile(`(^test_metrics.pb$|^test_metrics[_-]+[[:word:]-]*[[:alnum:]]\.pb$)`)
+	summaryFileRE := regexp.MustCompile(`(^test_summary.pb$|^test_summary[_-]+[[:word:]-]*[[:alnum:]]\.pb$)`)
+	allRE := []*regexp.Regexp{metricsFileRE, summaryFileRE}
+
+	reader, err := zip.OpenReader(zipFile)
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	for _, file := range reader.File {
+		if file.FileInfo().IsDir() {
+			continue
+		}
+
+		//  Look for a matching file name using regular expression patterns.
+		match := ""
+		for _, re := range allRE {
+			match = re.FindString(filepath.Base(file.Name))
+			if len(match) > 0 {
+				break
+			}
+		}
+		if len(match) == 0 {
+			continue
+		}
+		fmt.Printf("Found output file to process: %s\n", file.Name)
+
+		f, err := file.Open()
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+
+		// Read entire file contents into a byte slice.
+		data, err := ioutil.ReadAll(f)
```

Comment:
We need to explicitly bound the metrics filesize, or come up with a method of streaming this from an io.Reader instead of forcing the whole thing to be loaded into memory (in case someone generates e.g. 1GB of metrics)